internal/controller/argocdcommitstatus_controller.go

```diff
@@ -0,0 +1,420 @@
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.19.1/pkg/reconcile
+func (r *ArgoCDCommitStatusReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+	logger.Info("Reconciling ArgoCDCommitStatus")
+	var argoCDCommitStatus promoterv1alpha1.ArgoCDCommitStatus
+	err := r.Get(ctx, req.NamespacedName, &argoCDCommitStatus, &client.GetOptions{})
+	if err != nil {
+		if k8s_errors.IsNotFound(err) {
+			logger.Info("ArgoCDCommitStatus not found")
+			return ctrl.Result{}, nil
+		}
+
+		logger.Error(err, "failed to get ArgoCDCommitStatus")
+		return ctrl.Result{}, fmt.Errorf("failed to get ArgoCDCommitStatus: %w", err)
+	}
+
+	ls, err := metav1.LabelSelectorAsSelector(argoCDCommitStatus.Spec.ApplicationSelector)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to parse label selector: %w", err)
+	}
+	// TODO: we should setup a field index and only list apps related to the currently reconciled app
+	var ulArgoCDApps unstructured.UnstructuredList
+	ulArgoCDApps.SetGroupVersionKind(gvk)
+	err = r.Client.List(ctx, &ulArgoCDApps, &client.ListOptions{
+		LabelSelector: ls,
+	})
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to list CommitStatus objects: %w", err)
+	}
+
+	gitAuthProvider, repositoryRef, err := r.getGitAuthProvider(ctx, argoCDCommitStatus)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to get git auth provider: %w", err)
+	}
+
+	groupedArgoCDApps, err := r.groupArgoCDApplicationsWithPhase(&argoCDCommitStatus, ulArgoCDApps)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to get ArgoCDApplication: %w", err)
+	}
+
+	for key, appsInEnvironment := range groupedArgoCDApps {
+		repo, targetBranch := key.repo, key.targetBranch
+
+		gitOperation, err := git.NewGitOperations(ctx, r.Client, gitAuthProvider, r.PathLookup, repositoryRef, &argoCDCommitStatus, targetBranch)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to initialize git client: %w", err)
+		}
+
+		var resolvedSha string
+		resolvedSha, err = gitOperation.LsRemote(ctx, targetBranch)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to ls-remote sha: %w", err)
+		}
+
+		resolvedPhase, desc := r.calculateAggregatedPhaseAdnDescription(appsInEnvironment, resolvedSha)
+
+		mostRecentLastTransitionTime := r.getMostRecentLastTransitionTime(appsInEnvironment)
+
+		// Did the mostRecentLastTransitionTime occur more than 5 seconds ago
+		if mostRecentLastTransitionTime != nil && time.Since(mostRecentLastTransitionTime.Time) < 5*time.Second {
+			err = r.updateAggregatedCommitStatus(ctx, argoCDCommitStatus, targetBranch, repo, resolvedSha, promoterv1alpha1.CommitPhasePending, desc)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
```

Comment:
I think you could pass `mostRecentLastTransitionTime` to `calculateAggregatedPhaseAdnDescription` and have it handle setting `resolvedPhase` to `Pending` if the transition is too recent.

That would let you get rid of the `if` block and just call updateAggregatedCommitStatus` in one place.